pkg/controllers/management/multiclusterapp/controller_multiclusterapp.go

```diff
@@ -31,53 +31,52 @@ import (
+	clusterManager                *clustermanager.Manager
 	clusterLister                 v3.ClusterLister
 	projectLister                 v3.ProjectLister
-	clusterName                   string
 	userManager                   user.Manager
+	ctx                           context.Context
 }
 
-func Register(ctx context.Context, cluster *config.UserContext) {
-	mcApps := cluster.Management.Management.MultiClusterApps("")
-	m := &MCAppController{
-		apps:                          cluster.Management.Project.Apps(""),
-		appLister:                     cluster.Management.Project.Apps("").Controller().Lister(),
-		namespaces:                    cluster.Core.Namespaces(""),
+func StartMCAppManagementController(ctx context.Context, mgmt *config.ManagementContext, clusterManager *clustermanager.Manager) {
+	management := mgmt.Management
+	mcApps := management.MultiClusterApps("")
+	m := &MCAppManager{
+		ctx:                           ctx,
+		apps:                          mgmt.Project.Apps(""),
+		appLister:                     mgmt.Project.Apps("").Controller().Lister(),
 		multiClusterApps:              mcApps,
-		multiClusterAppRevisions:      cluster.Management.Management.MultiClusterAppRevisions(""),
-		multiClusterAppRevisionLister: cluster.Management.Management.MultiClusterAppRevisions("").Controller().Lister(),
-		nsLister:                      cluster.Core.Namespaces("").Controller().Lister(),
-		clusterLister:                 cluster.Management.Management.Clusters("").Controller().Lister(),
-		projectLister:                 cluster.Management.Management.Projects("").Controller().Lister(),
-		clusterName:                   cluster.ClusterName,
-		templateVersionLister:         cluster.Management.Management.CatalogTemplateVersions("").Controller().Lister(),
-		userManager:                   cluster.Management.UserManager,
-	}
+		multiClusterAppRevisions:      management.MultiClusterAppRevisions(""),
+		multiClusterAppRevisionLister: management.MultiClusterAppRevisions("").Controller().Lister(),
+		clusterManager:                clusterManager,
+		clusterLister:                 management.Clusters("").Controller().Lister(),
+		projectLister:                 management.Projects("").Controller().Lister(),
+		templateVersionLister:         management.CatalogTemplateVersions("").Controller().Lister(),
+		userManager:                   mgmt.UserManager,
+	}
+	mcAppTickerData = map[string]*IntervalData{}
 	m.multiClusterApps.AddHandler(ctx, "multi-cluster-app-controller", m.sync)
-
-	StartMCAppStateController(ctx, cluster)
 }
 
-func (m *MCAppController) sync(key string, mcapp *v3.MultiClusterApp) (runtime.Object, error) {
+func (m *MCAppManager) sync(key string, mcapp *v3.MultiClusterApp) (runtime.Object, error) {
 	if mcapp == nil || mcapp.DeletionTimestamp != nil {
 		_, mcappName, err := cache.SplitMetaNamespaceKey(key)
 		if err != nil {
 			return mcapp, err
 		}
+		deleteContext(mcappName)
```

Comment:
we should consider using cache with an expiry in case sync gets dropped for th e MCA

Reply:
opened issue to keep track `https://github.com/rancher/rancher/issues/18189`